File: mpi/hello.ex1.c
/* 
 * hello.ex1.c 
 * Parallel version using MPI calls
 * Modified from basic version so that workers send back 
 * a message to the master, who prints out a message 
 * for each worker 
 * RLF 10/11/95 
*/

#include <stdio.h>
#include <string.h>
#include "mpi.h"
main(int argc, char **argv )
{
  char message[20];
  int i,rank, worker, size, type=99;
  MPI_Status status;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD,&size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  if(rank == 0) {
    strcpy(message, "Hello, world");
    for (i=1; i<size; i++) 
      MPI_Send(message, 13, MPI_CHAR, i, type, MPI_COMM_WORLD);
    printf("process %d : %.13s\n", rank, message);
    for (i=1; i<size; i++) {
      MPI_Recv(&worker, 1, MPI_INT,MPI_ANY_SOURCE,type, 
               MPI_COMM_WORLD, &status);                 
      printf("process %d : Hello back\n", worker);
    }
  } 

  else {
    MPI_Recv(message, 20, MPI_CHAR, 0, type, MPI_COMM_WORLD, &status);
    MPI_Send(&rank, 1, MPI_INT, 0, type, MPI_COMM_WORLD);
  }

  MPI_Finalize();
}

File: mpi/hello.ex1.f
! Program hello.ex1.f
! Parallel version using MPI calls
! Modified from basic version so that workers send back
! a message to the master, who prints out a message
! for each worker
program hello
  implicit none
  integer, parameter:: DOUBLE=kind(1.0d0), SINGLE=kind(1.0)
  include "mpif.h"
  character(LEN=12) :: inmsg,message
  integer :: i,mpierr,rank,size,tag,wrank
  integer, dimension(MPI_STATUS_SIZE) :: status
  !
  call MPI_Init(mpierr)
  call MPI_Comm_size(MPI_COMM_WORLD,size,mpierr)
  call MPI_Comm_rank(MPI_COMM_WORLD,rank,mpierr)
  tag = 100
  !
  if (rank == 0) then
    message = "Hello, world"
    do i = 1,size-1
      call MPI_Send(message,12,MPI_CHARACTER,i,tag,MPI_COMM_WORLD,mpierr)
    end do
    write(*,*) "process", rank, ":", message
    do i = 1,size-1
      call MPI_Recv(wrank,1,MPI_INTEGER,MPI_ANY_SOURCE,tag,MPI_COMM_WORLD, &
                    status,mpierr)
      write(*,*) "process", wrank, ":Hello, back"
    end do
  else
    call MPI_Recv(inmsg,12,MPI_CHARACTER,0,tag,MPI_COMM_WORLD,status,mpierr)
    call MPI_Send(rank,1,MPI_INTEGER,0,tag,MPI_COMM_WORLD,mpierr)
  end if
  call MPI_Finalize(mpierr)
end program hello
!
